traffic_ops/app/lib/API/CachegroupFallback.pm

```diff
@@ -0,0 +1,208 @@
+	if ( !defined($cachegroup) ) {
+		&log( $self, "No such Cachegroup $id");
+		return $self->not_found();
+	}
+
+	if ( ($cachegroup->type->name ne "EDGE_LOC") ) {
+		return $self->alert("cachegroup should be type EDGE_LOC.");
+	}
+
+	if ( defined ($cache_id) ) {
+		my $rs_backups = $self->db->resultset('CachegroupFallback')->search({ primary_cg => $cache_id}, {order_by => 'set_order'});
+			if ( ($rs_backups->count > 0) ) {
+				my $response;
+				my $backup_cnt = 0;
+				while ( my $row = $rs_backups->next ) {
+					$response->[$backup_cnt]{"cacheGroupId"} = $cache_id;
+					$response->[$backup_cnt]{"cacheGroupName"} = $row->primary_cg->name;
+					$response->[$backup_cnt]{"fallbackName"} = $row->backup_cg->name;
+					$response->[$backup_cnt]{"fallbackId"} = $row->backup_cg->id;
+					$response->[$backup_cnt]{"fallbackOrder"} = $row->set_order;
+					$backup_cnt++;
+				}
+				return $self->success( $response );
+			} else {
+				&log( $self, "No backup Cachegroups for $cache_id");
+				return $self->not_found();
+			}
+
+	} elsif ( defined ($fallback_id) ) {
+		my $rs_backups = $self->db->resultset('CachegroupFallback')->search({ backup_cg => $fallback_id}, {order_by => 'set_order'});
+			if ( ($rs_backups->count > 0) ) {
+				my $response;
+				my $backup_cnt = 0;
+				while ( my $row = $rs_backups->next ) {
+					$response->[$backup_cnt]{"fallbackId"} = $fallback_id;
+					$response->[$backup_cnt]{"fallbackName"} = $row->backup_cg->name;
+					$response->[$backup_cnt]{"cacheGroupName"} = $row->primary_cg->name;
+					$response->[$backup_cnt]{"cacheGroupId"} = $row->primary_cg->id;
+					$response->[$backup_cnt]{"fallbackOrder"} = $row->set_order;
+					$backup_cnt++;
+				}
+				return $self->success( $response );
+			} else {
+				&log( $self, "No Cachegroups has $fallback_id as backup");
+				return $self->not_found();
+			}
+	}
+}
+
+sub update {
+	my $self = shift;
+	my $cache_id = $self->param('cacheGroupId');
+	my $params = $self->req->json;
+
+	if ( !defined($cache_id)) {
+		my @param_array = @{$params};
+		$cache_id = $param_array[0]{cacheGroupId};
+	}
+
+	if ( !defined($params) ) {
+		return $self->alert("parameters must be in JSON format,  please check!");
+	}
+
+	if ( !&is_oper($self) ) {
+		return $self->forbidden();
+	}
+
+	#only integers
+	if ( $cache_id !~ /^\d+?$/ ) {
+		&log( $self, "No such Cachegroup id $cache_id");
+		return $self->not_found();
+	}
+
+	my $cachegroup = $self->db->resultset('Cachegroup')->search( { id => $cache_id } )->single();
+	if ( !defined($cachegroup) ) {
+		return $self->not_found();
+	}
+
+	if ( ($cachegroup->type->name ne "EDGE_LOC") ) {
+		return $self->alert("cachegroup should be type EDGE_LOC.");
+	}
+
+	my $rs_backups = $self->db->resultset('CachegroupFallback')->search( { primary_cg => $cache_id } );
+	if ( defined ($rs_backups->next) ) {
+		&log( $self, "Deleting all existing backup configuration for cache group $cache_id");
+		$rs_backups->delete();
```

Comment:
this is deleting the entire list and replacing it with the contents of the POST

i thought we discussed simply adding the item if it doesn't exist in the db and updating it if it does.